### ISSUE TITLE
Job DSL configuration failing with "No signature of method DescribableListContext.bitBucketTrigger() is applicable for argument types"

### ISSUE THREAD

**Issue by tuukkamustonen** [jenkinsci/bitbucket-push-and-pull-request-plugin#166]
I have the plugin 2.7.0 with Job DSL 1.77 on Jenkins 2.249.2 and in a freestyle job:

```groovy
job('foobar') {
    properties {
        pipelineTriggers {
            triggers {
                bitBucketTrigger {
                    triggers {
                        bitBucketPPRPullRequestServerTriggerFilter {
                            actionFilter {
                                bitBucketPPRPullRequestServerCreatedActionFilter {
                                }
                            }
                        }
                        bitBucketPPRPullRequestServerTriggerFilter {
                            actionFilter {
                                bitBucketPPRPullRequestServerUpdatedActionFilter {
                                }
                            }
                        }
                        bitBucketPPRPullRequestServerTriggerFilter {
                            actionFilter {
                                bitBucketPPRPullRequestServerMergedActionFilter {
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    ...
}
```

But I get:

```
No signature of method: javaposse.jobdsl.plugin.structs.DescribableListContext.bitBucketTrigger()
is applicable for argument types: (pullrequests$_run_closure1$_closure4$_closure14$_closure15$_closure16) values:
[pullrequests$_run_closure1$_closure4$_closure14$_closure15$_closure16@59cd2c3b]
```

I've tried a few other ways (for example, with only one trigger), but I always get this sort of an error. The output isn't very helpful, and I'm not sure what I'm doing wrong.

I've read the docs (1.77+ syntax) and https://github.com/jenkinsci/bitbucket-push-and-pull-request-plugin/pull/160/files (which adds support for the new syntax). I'm fairly sure I don't have typos anywhere. I've also checked that the plugin is actually enabled and recognized by Job DSL plugin by checking `$JENKINS_URL/plugin/job-dsl/api-viewer/index.html`. Configuration via UI works just fine.

Any pointers?

**Comment by tuukkamustonen** [comment]
The old format:

```groovy
job('foobar') {
    triggers {
        bitbucketTriggers {
            pullRequestServerCreatedAction()
            pullRequestServerUpdatedAction()
            pullRequestServerMergedAction()
        }
    }
    ...
}
```

works flawlessly :O Why doesn't the new one?

**Comment by rhotau** [comment]
Hi @tuukkamustonen, would you by any chance have the Jenkins pipeline plugin installed (workflow-job is required, although I agree for freestyle jobs it seems counter-intuitive..)?

**Comment by tuukkamustonen** [comment]
Hi, sorry for extermely late reply (having troubles with my laptop).

If by "Jenknis pipeline plugin" you mean ẁorkflow-job` (so `https://plugins.jenkins.io/workflow-job/) then yes, that's installed, at `2.4.0` version.

**Comment by rhotau** [comment]
Unfortunately I do not have access to a BitbucketServer instance so I can only guess (I'm on a cloud/pipeline design which is the exact opposite of your setup..). If you set all parameters (although they shall be optional!) in the closure, for instance:
```
bitBucketPPRPullRequestServerCreatedActionFilter {
    allowedBranches("*")
    isToApprove(false)
}
```
does the error go away?

Another shot in the dark: if you use `freeStyleJob('foobar')` as the root instead of `job('foobar')`, does it work?

**Comment by tuukkamustonen** [comment]
Hey, I'll try these tomorrow. Will report back!

**Comment by tuukkamustonen** [comment]
I tried both ways, but no luck. Fails similarly as before.

> Unfortunately I do not have access to a BitbucketServer instance so I can only guess

You shouldn't need Bitbucket Server to replicate this - it happens during Job DSL loading, when the job is being built, so merely having a Jenkins instance should do.

**Comment by rhotau** [comment]
HI @tuukkamustonen, I'm a bit gobsmacked.. I did further experiments with a rather minimalistic seed job:
```
pipelineJob('foobar2') {
  properties {
    pipelineTriggers {
      triggers {
        bitBucketTrigger {
          triggers {
            bitBucketPPRPullRequestServerTriggerFilter {
              actionFilter {
                bitBucketPPRPullRequestServerCreatedActionFilter {
                }
              }
            }
            bitBucketPPRPullRequestServerTriggerFilter {
              actionFilter {
                bitBucketPPRPullRequestServerUpdatedActionFilter {
                }
              }
            }
            bitBucketPPRPullRequestServerTriggerFilter {
              actionFilter {
                bitBucketPPRPullRequestServerMergedActionFilter {
                }
              }
            }
          }
        }
      }
    }
  }
}
job('foobar') {
  properties {
    pipelineTriggers {
      triggers {
        bitBucketTrigger {
          triggers {
            bitBucketPPRPullRequestServerTriggerFilter {
              actionFilter {
                bitBucketPPRPullRequestServerCreatedActionFilter {
                }
              }
            }
            bitBucketPPRPullRequestServerTriggerFilter {
              actionFilter {
                bitBucketPPRPullRequestServerUpdatedActionFilter {
                }
              }
            }
            bitBucketPPRPullRequestServerTriggerFilter {
              actionFilter {
                bitBucketPPRPullRequestServerMergedActionFilter {
                }
              }
            }
          }
        }
      }
    }
  }
  steps {
    shell('echo START')
  }
}
```
I don't see errors during generation on my side but the freestyle job doesn't have the triggers set correctly (in fact it has none) while the pipeline-job has the correct triggers. This is beyond my knowledge but maybe freestyle job being older designs are not fully tailored for the fully-dynamical approach of pipeline/job-dsl that Jenkins core seems to be taking..
Regarding your issue, I would stick with what works for you (ie. the deprecated extension that shouldn't be remove unless job-dsl removes it as well) because I honestly wouldn't be able to dig in the dsl plugin code to understand what's going on..
As a sanity check, would you be able to run the above snippet (eventually without the `job('foobar')` that should fail) on your setup to see if the pipeline one gets generated correctly?
My plugins version are same as you: `bitbucket-push-and-pull-request:2.7.0`, `job-dsl:1.77`
I'm running with the groovy sandbox ticked although I doubt it could make much difference.
Sorry for not being more helpful 😞

**Comment by tuukkamustonen** [comment]
No worries, you're already being very helpful just looking into it, so thank you. I'll run those snippets tomorrow, will report back!

**Comment by tuukkamustonen** [comment]
Well... I can reproduce your behavior: Both snippets pass successfully, but only pipelineJob actually gets the plugin configured (checked via UI). In freestyle job there are no triggers even though Job DSL import passes.

I tried to reproduce my previous problem, where I got the closure error, but it's now gone :O I haven't updated plugins since, but I did restart Jenkins at some point. Maybe there was something pending a restart - I may have installed your plugin without restart, but maybe it requires one?

A note to self: always restart Jenkins in scenarios like this.

So, the situation is this:

- In pipeline job, 1.77+ job DSL format works.
- In freestyle job, 1.77+ job DSL format passes silently, but triggers don't really appear to the job.
- In freestyle job, pre-1.77 job DSL format works.

I guess users can live using the deprecated Job DSL syntax, but I fear your going to get more reports about the new syntax not working, should someone attempt to use the plugin for freestyle jobs, too.

**Comment by tuukkamustonen** [comment]
Oh, I also noticed that this option seems to be completely unsupported in Job DSL (any version):

![image](https://user-images.githubusercontent.com/94327/109944590-0c3fef80-7cdf-11eb-9f9e-1a52d6a7f3d0.png)

Should it be there?

**Comment by tuukkamustonen** [comment]
Turns out it is supported via:

    pullRequestServerSourceUpdatedAction
    bitBucketPPRPullRequestServerSourceUpdatedActionFilter

But it's undocumented at https://plugins.jenkins.io/bitbucket-push-and-pull-request/

**Comment by rhotau** [comment]
Hi @tuukkamustonen, I'm wondering if it isn't just down to a misalignment between UI/freestyle/pipeline.. I'm attaching a gist containing the config.xml file I retrieved from the server after generating `job('foobar')` from the above snippet, and I can see a trigger 😮, however it is linked to workflow (old name for pipeline) and not freestyle property: `<org.jenkinsci.plugins.workflow.job.properties.PipelineTriggersJobProperty plugin="workflow-job@2.40">`.
Extracting a freestyle job with trigger from same server, I'm seeing that it gets populated as:
```
<project>
  ...
  <triggers>
    <io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRTrigger plugin="bitbucket-push-and-pull-request@2.7.0">
      <spec></spec>
      <triggers>
        <io.jenkins.plugins.bitbucketpushandpullrequest.filter.pullrequest.server.BitBucketPPRPullRequestServerTriggerFilter>
          <actionFilter class="io.jenkins.plugins.bitbucketpushandpullrequest.filter.pullrequest.server.BitBucketPPRPullRequestServerApprovedActionFilter">
            <allowedBranches></allowedBranches>
            <triggerOnlyIfAllReviewersApproved>false</triggerOnlyIfAllReviewersApproved>
            <isToApprove>false</isToApprove>
          </actionFilter>
        </io.jenkins.plugins.bitbucketpushandpullrequest.filter.pullrequest.server.BitBucketPPRPullRequestServerTriggerFilter>
      </triggers>
    </io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRTrigger>
  </triggers>
  ...
</project>
```
and in this case, the UI is showing it as you would expect.
It seems that the job-dsl is correctly doing what it is told to, however at this point I couldn't tell whether it would work if such trigger occur. Have you tried whether the trigger works on your side? If so it could just be a documentation update to inform about this UI misalignment.

Regarding your second question, I think it was reported in #143.

gist: https://gist.github.com/rhotau/be7426d458b6fd5740b2546d018e4ad5

**Comment by rhotau** [comment]
Finally wrapped my head around it this morning.. In fact the `triggers()` context has only been deprecated in jod-dsl for pipeline jobs, freestyle jobs are still relying on it. The deprecated annotation should not have been added as part of this plugin (I read the wiki too quickly and it bit me). My apologies for causing disruptions in the first place 😞.

**Comment by tuukkamustonen** [comment]
I checked it.

Creating a job via Job DSL with:

    job('...')
    freeStyleJob('...')
    pipeline('...')

For all of these, the `config.xml` gets:

```xml
  <properties>
    <org.jenkinsci.plugins.workflow.job.properties.PipelineTriggersJobProperty plugin="workflow-job@2.40">
    ...
```

So it uses workflow-property like you said.

But it's kinda weird, as pipeline job shows it correctly in UI while `job` and `freeStyleJob` don't.

I couldn't check what gets set if I set/adjust it via UI as our Jenkins server broke down today due to (some other) reason :D
